### PR TITLE
VZ-10525: Delete Keycloak StatefulSet during upgrade only when required

### PR DESCRIPTION
**Pull request by bhatpmk**
Delete Keycloak StatefulSet only during Verrazzano upgrade from pre 1.5.0 to 1.5.0 and onwards. The deletion of StatefulSet is required only when upgrading Keycloak 15.0.2 to 20.0.1, as the Helm chart for Keycloak 20.0.1 contains changes to fields other than 'replicas', 'template', and 'updateStrategy'.

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go [hunk 1]
```diff
@@ -254,3 +263,34 @@ func (c KeycloakComponent) MonitorOverrides(ctx spi.ComponentContext) bool {
...
+
+// isDeleteStatefulSetRequired determines whether deleting the Keycloak StatefulSet is required
+func isDeleteStatefulSetRequired(ctx spi.ComponentContext) (bool, error) {
+	isDeleteRequired := true
+	if ctx.ActualCR() != nil {
+		installedVersion, err := semver.NewSemVersion(ctx.ActualCR().Status.Version)
+		if err != nil {
+			return isDeleteRequired, nil
+		}
+		ctx.Log().Infof("Verrazzano version installed: %s", installedVersion)
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go:275]: please make this debug

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go [hunk 2]
```diff
@@ -254,3 +263,34 @@ func (c KeycloakComponent) MonitorOverrides(ctx spi.ComponentContext) bool {
 	}
 	return false
 }
+
+// isDeleteStatefulSetRequired determines whether deleting the Keycloak StatefulSet is required
+func isDeleteStatefulSetRequired(ctx spi.ComponentContext) (bool, error) {
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go:268]: can you add a comment to describe how you are determining if delete is required?

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go [hunk 3]
```diff
@@ -254,3 +263,34 @@ func (c KeycloakComponent) MonitorOverrides(ctx spi.ComponentContext) bool {
...
+		minVersion, err := semver.NewSemVersion(constants.VerrazzanoVersion1_5_0)
+		if err != nil {
+			return isDeleteRequired, nil
+		}
+
+		// Deleting the Keycloak StatefulSet during pre-upgrade is required only when the version installed is pre 1.5.0
+		// (with Keycloak version 15.0.2) and version to upgrade is 1.5.0 or later(with Keycloak version 20.0.1)
+		if (installedVersion.IsLessThan(minVersion) && versionToUpgrade.IsLessThan(minVersion)) ||
+			!installedVersion.IsLessThan(minVersion) {
+			ctx.Log().Info("Deleting Keycloak StatefulSet is not during pre-upgrade")
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go:291]: log oncef

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go (hunk 2 above)
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go:268]: better to put this function in keycloak.go

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go [hunk 4]
```diff
@@ -254,3 +263,34 @@ func (c KeycloakComponent) MonitorOverrides(ctx spi.ComponentContext) bool {
...
 	return false
 }
+
+// isDeleteStatefulSetRequired determines whether deleting the Keycloak StatefulSet is required
+func isDeleteStatefulSetRequired(ctx spi.ComponentContext) (bool, error) {
+	isDeleteRequired := true
+	if ctx.ActualCR() != nil {
+		installedVersion, err := semver.NewSemVersion(ctx.ActualCR().Status.Version)
+		if err != nil {
+			return isDeleteRequired, nil
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go:273]: so this is saying, if we can't determine the version, go ahead and do the delete anyways, correct?  Seems right, just making sure.

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go [hunk 5]
```diff
@@ -254,3 +263,34 @@ func (c KeycloakComponent) MonitorOverrides(ctx spi.ComponentContext) bool {
...
+		ctx.Log().Infof("Verrazzano version for upgrade: %s", versionToUpgrade)
+
+		minVersion, err := semver.NewSemVersion(constants.VerrazzanoVersion1_5_0)
+		if err != nil {
+			return isDeleteRequired, nil
+		}
+
+		// Deleting the Keycloak StatefulSet during pre-upgrade is required only when the version installed is pre 1.5.0
+		// (with Keycloak version 15.0.2) and version to upgrade is 1.5.0 or later(with Keycloak version 20.0.1)
+		if (installedVersion.IsLessThan(minVersion) && versionToUpgrade.IsLessThan(minVersion)) ||
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go:289]: don't you just need to know that the installed version is >= 1.5.0?  this method already knows we're in pre-upgrade, and we just need to know that the installed version has already handled the Keycloak 15 --> 20 upgrade.

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go (hunk 3 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go:291]: message is awkward, maybe `Deleting Keycloak Statefulset is not required during pre-upgrade`.  Think it just is missing the word `required`.

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go (hunk 5 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go:289]: Probably could be simplified to

```
if versionToUpgrade.IsLessThan(minVersion) || !installedVersion.IsLessThan(minVersion) {
...
}
```

Comment (bhatpmk) [reply platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go:289]: We don't need to delete the StatefulSet when the upgrade if from - 1.4.x to 1.4.y

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go (hunk 4 above)
Comment (bhatpmk) [reply platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go:273]: Yes, that is the intention. Deleting the StatefulSet won't harm, although it is not necessary in some cases, as handled by this change

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go (hunk 3 above)
Comment (bhatpmk) [reply platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go]: Yes, missed "required"

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go (hunk 2 above)
Comment (bhatpmk) [reply platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go]: Done

File: platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go (hunk 1 above)
Comment (bhatpmk) [reply platform-operator/controllers/verrazzano/component/keycloak/keycloak_component.go]: Done. Actually I wanted to remove it before creating the PR, forgot it completely

**Review by michael-cico — APPROVED** [review]

**Comment by bhatpmk** [comment]
VZ-10525

**Pull request merged**